pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -423,102 +425,108 @@ func (a Adapter) waitAndGetComponentPod(hideSpinner bool) (*corev1.Pod, error) {
 	return pod, nil
 }
 
-// Push syncs source code from the user's disk to the component
+// Executes Devfile Commands
 func (a Adapter) execDevfile(pushDevfileCommands []versionsCommon.DevfileCommand, componentExists, show bool, podName string, containers []corev1.Container) (err error) {
-	var buildRequired bool
 	var s *log.Status
 
-	if len(pushDevfileCommands) == 1 {
-		// if there is one command, it is the mandatory run command. No need to build.
-		buildRequired = false
-	} else if len(pushDevfileCommands) == 2 {
-		// if there are two commands, it is the optional build command and the mandatory run command, set buildRequired to true
-		buildRequired = true
-	} else {
-		return fmt.Errorf("error executing devfile commands - there should be at least 1 command or at most 2 commands, currently there are %v commands", len(pushDevfileCommands))
+	type CommandNames struct {
+		defaultName string
+		adapterName string
 	}
 
-	for i := 0; i < len(pushDevfileCommands); i++ {
-		command := pushDevfileCommands[i]
-
-		// Exec the devBuild command if buildRequired is true
-		if (command.Name == string(common.DefaultDevfileBuildCommand) || command.Name == a.devfileBuildCmd) && buildRequired {
-			glog.V(3).Infof("Executing devfile command %v", command.Name)
-
-			for _, action := range command.Actions {
-				// Change to the workdir and execute the command
-				var cmdArr []string
-				if action.Workdir != nil {
-					cmdArr = []string{"/bin/sh", "-c", "cd " + *action.Workdir + " && " + *action.Command}
-				} else {
-					cmdArr = []string{"/bin/sh", "-c", *action.Command}
-				}
-
-				if show {
-					s = log.SpinnerNoSpin("Executing " + command.Name + " command " + fmt.Sprintf("%q", *action.Command))
+	commandOrder := []CommandNames{
+		CommandNames{defaultName: string(common.DefaultDevfileInitCommand), adapterName: a.devfileInitCmd},
+		CommandNames{defaultName: string(common.DefaultDevfileBuildCommand), adapterName: a.devfileBuildCmd},
+		CommandNames{defaultName: string(common.DefaultDevfileRunCommand), adapterName: a.devfileRunCmd},
+	}
+	for i, currentCommand := range commandOrder {
+		for _, command := range pushDevfileCommands {
+			if command.Name == currentCommand.defaultName || command.Name == currentCommand.adapterName {
```

Comment:
We no longer need the `buildRequired` bool anymore because we verify that the commands are required in `ValidateAndGetPushDevfileCommands` - an `init` can't be provided without a `build`, so we don't add it to the array of commands, and like the old code we currently make the assumption that if only one command is given it is `run`. This means the only possible combinations of `pushDevfileCommands` are: [init build run], [build run], and [run]. Since this function is internal and only gets called once after we retrieve the commands, it should be fine. However, if this function would be called in a different scenario, an additional check would be required.

Reply:
On a related note: should we error out if the `pushDevfileCommands` slice is empty?

Reply:
the validate function should handle that, you can try it out with your PR to double confirm

Reply:
You're right, it does look like the Validate function checks that the run command exists, I can add an additional check in the  `execDevfile` function too for good measure though.